File: nginx/fake_graceful_nginx.py
#!/usr/bin/env python3.6

import signal
import sys
import time

def sigquit_handler(sig, frame):
    time.sleep(0.5)
    print('Received sigquit, doing graceful shutdown')
    sys.exit(0)

# Can't do anything in this handler - python libs are not thread safe, so not safe to call e.g. print.
def sighup_handler(sig, frame):
    pass

print('Running {}'.format(str(sys.argv)))

if sys.argv[1] == '-v':
    print('Asked for version')
    sys.exit(0)

if sys.argv[1] == '-t':
    print('Asked for config validation')
    sys.exit(0)

# The parent golang process blocks SIGQUIT in subprocesses, for some reason.
# So we unblock it manually - same as what nginx does.
signal.pthread_sigmask(signal.SIG_UNBLOCK, {signal.SIGQUIT, signal.SIGHUP})
signal.signal(signal.SIGQUIT, sigquit_handler)
signal.signal(signal.SIGHUP, sighup_handler)
signal.pause

startup_marker_file_name = str.join('/', sys.argv[2].split('/')[:-1]) + '/nginx-started'
with open(startup_marker_file_name, 'w') as f:
    f.write('started!')

time.sleep(5)
print('Quit after 5 seconds of nada')
sys.exit(-1)

File: nginx/fake/fake_graceful_nginx.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"os/signal"
	"path"
	"strings"
	"syscall"
	"time"
)

func main() {
	sigCh := make(chan os.Signal, 1)
	signal.Notify(sigCh, syscall.SIGQUIT, syscall.SIGHUP)

	if os.Args[1] == "-v" {
		fmt.Println("Asked for version")
		os.Exit(0)
	}

	if os.Args[1] == "-t" {
		fmt.Println("Asked for config validation")
		os.Exit(0)
	}

	startupMarkerFilename := startupMarkerFilename()
	err := ioutil.WriteFile(startupMarkerFilename, []byte("started!"), os.ModePerm)
	if err != nil {
		panic(err)
	}

	timer := time.NewTimer(5 * time.Second)

	select {
	case sig := <-sigCh:
		if sig == syscall.SIGQUIT {
			time.Sleep(500 * time.Millisecond)
			fmt.Println("Received sigquit, doing graceful shutdown")
			os.Exit(0)
		}

		if sig == syscall.SIGHUP {
			err := ioutil.WriteFile(startupMarkerFilename, []byte("reloaded!"), os.ModePerm)
			if err != nil {
				panic(err)
			}
		}

	case <-timer.C:
		fmt.Println("Quit after 5 seconds of nada")
		os.Exit(-1)
	}
}

func startupMarkerFilename() string {
	filename := strings.Split(os.Args[2], "/")
	filename = filename[:len(filename)-1]
	filename = append(filename, "nginx-log")
	return "/" + path.Join(filename...)
}
